routing.go

```diff
@@ -500,132 +421,154 @@ func (dht *IpfsDHT) FindProviders(ctx context.Context, c cid.Cid) ([]peer.AddrIn
+// FindProvidersAsyncExtended searches until the context expires.
+func (dht *IpfsDHT) FindProvidersAsyncExtended(ctx context.Context, key cid.Cid, opts ...routing.Option) (<-chan peer.AddrInfo, <-chan []peer.ID, error) {
 	if !dht.enableProviders || !key.Defined() {
-		peerOut := make(chan peer.AddrInfo)
+		peerOut, closestPeers := make(chan peer.AddrInfo), make(chan []peer.ID)
 		close(peerOut)
-		return peerOut
+		close(closestPeers)
+		return peerOut, closestPeers, routing.ErrNotSupported
 	}
 
-	chSize := count
-	if count == 0 {
-		chSize = 1
+	var cfg routing.Options
+	if err := cfg.Apply(opts...); err != nil {
+		return nil, nil, err
 	}
-	peerOut := make(chan peer.AddrInfo, chSize)
 
 	keyMH := key.Hash()
 
-	go dht.findProvidersAsyncRoutine(ctx, keyMH, count, peerOut)
-	return peerOut
+	processors := dhtrouting.GetProcessors(&cfg)
+	if processors == nil {
+		newValuesOnly := &dhtrouting.NewPeerIDFilter{
+			Key:   string(keyMH),
+			Peers: peer.NewSet(),
+		}
+
+		quorum := &dhtrouting.CountStopper{
+			MaxCount: dhtrouting.GetQuorum(&cfg),
+		}
+
+		processors = []dhtrouting.Processor{newValuesOnly, quorum}
+	}
+
+	return dhtrouting.FindProviders(ctx, keyMH, dht.findProvidersAsyncRoutine, processors, &cfg)
 }
 
-func (dht *IpfsDHT) findProvidersAsyncRoutine(ctx context.Context, key multihash.Multihash, count int, peerOut chan peer.AddrInfo) {
+// FindProvidersAsync is the same thing as FindProviders, but returns a channel.
+// Peers will be returned on the channel as soon as they are found, even before
+// the search query completes. If count is zero then the query will run until it
+// completes. Note: not reading from the returned channel may block the query
+// from progressing.
+func (dht *IpfsDHT) FindProvidersAsync(ctx context.Context, key cid.Cid, count int) <-chan peer.AddrInfo {
+	providers, _, _ := dht.FindProvidersAsyncExtended(ctx, key, dhtrouting.Quorum(count))
+	return providers
+}
+
+func (dht *IpfsDHT) findProvidersAsyncRoutine(ctx, queryAbortedCtx context.Context, key multihash.Multihash, seedPeerOpts dhtrouting.SeedPeersOptions) (<-chan peer.AddrInfo, <-chan []peer.ID) {
 	logger.Debugw("finding providers", "key", key)
 
-	defer close(peerOut)
+	provsCh := make(chan peer.AddrInfo, 1)
+	closestPeersCh := make(chan []peer.ID, 1)
 
-	findAll := count == 0
-	var ps *peer.Set
-	if findAll {
-		ps = peer.NewSet()
-	} else {
-		ps = peer.NewLimitedSet(count)
-	}
+	go func() {
+		defer close(provsCh)
+		defer close(closestPeersCh)
 
-	provs := dht.ProviderManager.GetProviders(ctx, key)
-	for _, p := range provs {
-		// NOTE: Assuming that this list of peers is unique
-		if ps.TryAdd(p) {
+		provs := dht.ProviderManager.GetProviders(ctx, key)
+		for _, p := range provs {
 			pi := dht.peerstore.PeerInfo(p)
 			select {
-			case peerOut <- pi:
+			case provsCh <- pi:
 			case <-ctx.Done():
 				return
 			}
 		}
 
-		// If we have enough peers locally, don't bother with remote RPC
-		// TODO: is this a DOS vector?
```

Comment:
If people start querying with count = 0 then this isn't a problem, so may be we should just emphasize that?